keras/src/models/cloning.py

```diff
@@ -170,22 +171,53 @@ def call_function(layer, *args, **kwargs):
             cache=cache,
         )
 
+        if isinstance(model, CompositeLayer):
+            cloned_inputs, cloned_outputs = _clone_function_object(
+                model._function,
+                clone_function=clone_function,
+                call_function=call_function,
+                input_tensors=input_tensors,
+            )
+            # Create a Keras Function from the graph between inputs and outputs
+            function = Function(cloned_inputs, cloned_outputs,
+                                model._function.name)
+
+            # Create a new CompositeLayer from the cloned function
+            # Note: A functional subclass of CompositeLayer will be
+            #       cloned as a vanilla CompositeLayer. This could be changed
+            #       in the future to;
+            #       inst = layer.__class__.__new__, then
+            #       CompositeLayer.__init__(inst, function, layer.name)
+            #       It would represent the cloned CompositeLayer with 
+            #       the correct class name but not call the __init__
+            #       method of the subclass which could create problems.
+            return CompositeLayer(function, model.name)
+
         # If the get_config() method is the same as a regular Functional
-        # model, we're safe to use _clone_functional_model (which relies
+        # model, we're safe to use _clone_function_object (which relies
         # on a Functional constructor). In the case where the get_config
         # is custom, this may not necessarily work, but if clone_function
-        # or input_tensors are passed, we attempt it anyway
+        # or call_function or input_tensors are passed, we attempt it anyway
         # in order to preserve backwards compatibility.
-        if utils.is_default(model.get_config) or (
-            clone_function or input_tensors
-        ):
-            return _clone_functional_model(
-                model,
+        if (utils.is_default(model.get_config) or
+            (clone_function or call_function or input_tensors)
+            ):
+            cloned_inputs, cloned_outputs = _clone_function_object(
+                model, # the model is a Function
                 clone_function=clone_function,
                 call_function=call_function,
                 input_tensors=input_tensors,
             )
 
+            # A subclassed Functional model is always cloned
+            #  as a vanilla Functional model.
+            new_model = Functional(cloned_inputs, cloned_outputs,
+                                    name=model.name)
+            if model.compiled:
+                compiled_config = model.get_compile_config()
+                new_model.compile_from_config(compiled_config)
+            return new_model
+         
```

Comment:
This piece of code was moved here from the end of _clone_functional_model wjen the function was renames _clone_function_object and repurposed for all functional types.